File: compilers/zig/src/ir/n113_ir_trust_boundary_test.zig
//! N-113 — the two shapes the Go tier rejected alone (R-079 / R-081).
//!
//! Both are `--ir`-only: the source path refuses each shape in
//! `passes/validate.zig`, and `parseANFProgram` is reached only from the IR
//! loader. Go grew these guards first (compilers/go/ir/loader.go) and was
//! deliberately, transiently stricter than its six peers until N-113;
//! `conformance/negatives/ir-rejection-parity.test.ts` is the gate that now
//! compares the six.
//!
//! Zig's IR loader is an error-enum channel with no message payload, so the
//! error NAME is the whole diagnostic — hence `EmptyRawScriptBody` and
//! `NoPublicMethods` rather than another `InvalidConstValue`.

const std = @import("std");
const testing = std.testing;
const json = @import("json.zig");

/// A one-method contract parameterised on the two fields under test, so every
/// case below differs from the VALID control in exactly one way.
fn irJson(allocator: std.mem.Allocator, is_public: bool, raw_bytes: []const u8, out_arity: u8) ![]const u8 {
    return std.fmt.allocPrint(allocator,
        \\{{"contractName":"Anyone","properties":[],"methods":[{{"name":"unlock","params":[],"isPublic":{},"body":[{{"name":"t0","value":{{"kind":"raw_script","bytes":"{s}","in_arity":0,"out_arity":{d}}}}}]}}]}}
    , .{ is_public, raw_bytes, out_arity });
}

// The control every case below is derived from. A probe whose control also
// fails proves nothing.
test "control: valid IR is accepted" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const src = try irJson(allocator, true, "51", 1);
    const program = try json.parseANFProgram(allocator, src);
    try testing.expectEqualStrings("Anyone", program.contract_name);
}

// R-079: lowering pops in_arity and pushes out_arity on the stack model while
// emission writes nothing for a zero-length span. The span degrades to the
// identity function and a DIFFERENT WITNESS spends the output. Measured on
// @bsv/sdk's Spend: `8f01859c` accepts x=5 and rejects x=-5; with the body
// erased, `01859c` does the opposite.
test "rejects an empty raw_script body" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const src = try irJson(allocator, true, "", 1);
    try testing.expectError(error.EmptyRawScriptBody, json.parseANFProgram(allocator, src));
}

// The degenerate in=0/out=0 case is harmless on its own and is rejected
// anyway: mirroring the source validator exactly beats a narrower
// arity-conditional rule that would differ from the rule one pass earlier.
test "rejects an empty raw_script body even at zero arity" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const src = try irJson(allocator, true, "", 0);
    try testing.expectError(error.EmptyRawScriptBody, json.parseANFProgram(allocator, src));
}

// R-081: emission succeeds with an EMPTY locking script, which is
// anyone-can-spend. On @bsv/sdk's Spend under full consensus wrappers,
// lock="" with unlock=OP_1 (0x51) validates.
test "rejects a contract with no public methods" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const src = try irJson(allocator, false, "51", 1);
    try testing.expectError(error.NoPublicMethods, json.parseANFProgram(allocator, src));
}

test "rejects an empty method list" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const src =
        \\{"contractName":"Empty","properties":[],"methods":[]}
    ;
    try testing.expectError(error.NoPublicMethods, json.parseANFProgram(allocator, src));
}

// Ordering matters and is asserted, not assumed: when a binding is ALSO
// malformed, the malformed binding is the more actionable diagnostic. Same
// ordering as compilers/go/ir/loader.go — the per-binding parse runs first, so
// the odd-hex error wins over the missing entry point.
test "structural errors keep priority over the entry-point error" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const src = try irJson(allocator, false, "515", 1);
    try testing.expectError(error.InvalidConstValue, json.parseANFProgram(allocator, src));
}

File: compilers/java/src/test/java/runar/compiler/passes/N113IrTrustBoundaryTest.java
package runar.compiler.passes;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;
import runar.compiler.ir.anf.AnfProgram;

/**
 * N-113 — the two shapes the Go tier rejected alone (R-079 / R-081).
 *
 * <p>Both are {@code --ir}-only: the source path refuses each shape in
 * {@code Validate.java}, and {@code AnfLoader.parse} is reached only from the
 * {@code --ir} path. Go grew these guards first
 * ({@code compilers/go/ir/loader.go}) and was deliberately, transiently
 * stricter than its six peers until N-113;
 * {@code conformance/negatives/ir-rejection-parity.test.ts} is the gate that
 * now compares the six.
 */
class N113IrTrustBoundaryTest {

    /**
     * A one-method contract parameterised on the two fields under test, so
     * every case below differs from the VALID control in exactly one way.
     */
    private static String ir(boolean isPublic, String rawBytes, int outArity) {
        return """
            {"contractName":"Anyone","properties":[],"methods":[
              {"name":"unlock","params":[],"isPublic":%s,"body":[
                {"name":"t0","value":{"kind":"raw_script","bytes":"%s","in_arity":0,"out_arity":%d}}
              ]}
            ]}
            """.formatted(isPublic, rawBytes, outArity);
    }

    /** The control every case below is derived from. A probe whose control
     *  also fails proves nothing. */
    @Test
    void controlValidIrIsAccepted() {
        AnfProgram p = AnfLoader.parse(ir(true, "51", 1));
        assertEquals("Anyone", p.contractName());
    }

    /**
     * R-079: lowering pops in_arity and pushes out_arity on the stack model
     * while emission writes nothing for a zero-length span. The span degrades
     * to the identity function and a DIFFERENT WITNESS spends the output.
     * Measured on @bsv/sdk's Spend: {@code 8f01859c} accepts x=5 and rejects
     * x=-5; with the body erased, {@code 01859c} does the opposite.
     */
    @Test
    void rejectsEmptyRawScriptBody() {
        RuntimeException ex = assertThrows(
            RuntimeException.class, () -> AnfLoader.parse(ir(true, "", 1)));
        assertTrue(ex.getMessage().contains("empty bytes body"), ex.getMessage());
    }

    /**
     * The degenerate in=0/out=0 case is harmless on its own and is rejected
     * anyway: mirroring the source validator exactly beats a narrower
     * arity-conditional rule that would differ from the rule one pass earlier.
     */
    @Test
    void rejectsEmptyRawScriptBodyEvenAtZeroArity() {
        assertThrows(RuntimeException.class, () -> AnfLoader.parse(ir(true, "", 0)));
    }

    /**
     * R-081: emission succeeds with an EMPTY locking script, which is
     * anyone-can-spend. On @bsv/sdk's Spend under full consensus wrappers,
     * {@code lock=""} with {@code unlock=OP_1 (0x51)} validates.
     */
    @Test
    void rejectsNoPublicMethods() {
        RuntimeException ex = assertThrows(
            RuntimeException.class, () -> AnfLoader.parse(ir(false, "51", 1)));
        assertTrue(ex.getMessage().contains("no public methods"), ex.getMessage());
    }

    @Test
    void rejectsEmptyMethodList() {
        RuntimeException ex = assertThrows(RuntimeException.class, () ->
            AnfLoader.parse("{\"contractName\":\"Empty\",\"properties\":[],\"methods\":[]}"));
        assertTrue(ex.getMessage().contains("no public methods"), ex.getMessage());
    }

    /**
     * Ordering matters and is asserted, not assumed: when a binding is ALSO
     * malformed, the malformed binding is the more actionable diagnostic. Same
     * ordering as {@code compilers/go/ir/loader.go} — methods are decoded
     * before the entry-point check runs.
     */
    @Test
    void structuralErrorsKeepPriorityOverTheEntryPointError() {
        RuntimeException ex = assertThrows(
            RuntimeException.class, () -> AnfLoader.parse(ir(false, "515", 1)));
        assertTrue(ex.getMessage().contains("odd hex length"), ex.getMessage());
    }
}
